File: Linear Regression/Python_Version/fmincg.py
import numpy as np
import sys

def fmincg(f, X, options=None, *args):
    """
    Minimize a continuous differentiable multivariate function. Starting point
    is given by "X" (D by 1), and the function "f" must return a function value 
    and a vector of partial derivatives.
    
    Usage: X, fX, i = fmincg(f, X, options, P1, P2, P3, P4, P5)
    
    Parameters:
    -----------
    f : function
        Function to minimize. Should return (function_value, gradient)
    X : array-like
        Starting point (1D array)
    options : dict, optional
        Dictionary containing 'MaxIter' key for maximum iterations
    *args : variable args
        Additional arguments to pass to function f
    
    Returns:
    --------
    X : ndarray
        Optimized parameters
    fX : list
        Function values at each iteration
    i : int
        Number of iterations performed
    """
    
    # Read options
    if options is not None and 'MaxIter' in options:
        length = options['MaxIter']
    else:
        length = 100
    
    # Constants for line searches
    RHO = 0.01      # RHO and SIG are constants in the Wolfe-Powell conditions
    SIG = 0.5
    INT = 0.1       # Don't reevaluate within 0.1 of the limit
    EXT = 3.0       # Extrapolate maximum 3 times the current bracket
    MAX = 20        # Max 20 function evaluations per line search
    RATIO = 100     # Maximum allowed slope ratio
    
    # Handle length as [maxiter, reduction] pair
    if isinstance(length, (list, tuple, np.ndarray)):
        if len(length) == 2:
            red = length[1]
            length = length[0]
        else:
            red = 1
    else:
        red = 1
    
    # Initialize
    i = 0
    ls_failed = 0
    fX = []
    X = np.array(X, dtype=float).flatten()
    
    # Get initial function value and gradient
    f1, df1 = f(X, *args)
    df1 = np.array(df1, dtype=float).flatten()
    
    i = i + (1 if length < 0 else 0)
    s = -df1  # Search direction is steepest descent
    d1 = -np.dot(s, s)  # Slope
    z1 = red / (1 - d1)  # Initial step
    
    # Main optimization loop
    while i < abs(length):
        i = i + (1 if length > 0 else 0)
        
        X0 = X.copy()
        f0 = f1
        df0 = df1.copy()
        
        X = X + z1 * s  # Begin line search
        f2, df2 = f(X, *args)
        f2 = float(f2)
        df2 = np.array(df2, dtype=float).flatten()
        
        i = i + (1 if length < 0 else 0)
        d2 = np.dot(df2, s)
        f3 = f1
        d3 = d1
        z3 = -z1
        
        M = MAX if length > 0 else min(MAX, -length - i)
        success = 0
        limit = -1
        
        # Line search loop
        while True:
            # Inner loop for bracketing
            while (f2 > f1 + z1 * RHO * d1 or d2 > -SIG * d1) and M > 0:
                limit = z1
                
                if f2 > f1:
                    # Quadratic fit
                    z2 = z3 - (0.5 * d3 * z3 * z3) / (d3 * z3 + f2 - f3)
                else:
                    # Cubic fit
                    A = 6 * (f2 - f3) / z3 + 3 * (d2 + d3)
                    B = 3 * (f3 - f2) - z3 * (d3 + 2 * d2)
                    z2 = (np.sqrt(B * B - A * d2 * z3 * z3) - B) / A
                
                # Handle numerical errors
                if np.isnan(z2) or np.isinf(z2):
                    z2 = z3 / 2
                
                z2 = max(min(z2, INT * z3), (1 - INT) * z3)
                z1 = z1 + z2
                X = X + z2 * s
                f2, df2 = f(X, *args)
                f2 = float(f2)
                df2 = np.array(df2, dtype=float).flatten()
                M = M - 1
                i = i + (1 if length < 0 else 0)
                d2 = np.dot(df2, s)
                z3 = z3 - z2
            
            # Check line search result
            if f2 > f1 + z1 * RHO * d1 or d2 > -SIG * d1:
                break  # Failure
            elif d2 > SIG * d1:
                success = 1
                break  # Success
            elif M == 0:
                break  # Failure
            
            # Make cubic extrapolation
            A = 6 * (f2 - f3) / z3 + 3 * (d2 + d3)
            B = 3 * (f3 - f2) - z3 * (d3 + 2 * d2)
            discriminant = B * B - A * d2 * z3 * z3
            
            if discriminant >= 0:
                z2 = -d2 * z3 * z3 / (B + np.sqrt(discriminant))
            else:
                z2 = np.nan
            
            # Handle numerical problems
            if np.isnan(z2) or np.isinf(z2) or z2 < 0:
                if limit < -0.5:
                    z2 = z1 * (EXT - 1)
                else:
                    z2 = (limit - z1) / 2
            elif limit > -0.5 and z2 + z1 > limit:
                z2 = (limit - z1) / 2
            elif limit < -0.5 and z2 + z1 > z1 * EXT:
                z2 = z1 * (EXT - 1.0)
            elif z2 < -z3 * INT:
                z2 = -z3 * INT
            elif limit > -0.5 and z2 < (limit - z1) * (1.0 - INT):
                z2 = (limit - z1) * (1.0 - INT)
            
            f3 = f2
            d3 = d2
            z3 = -z2
            z1 = z1 + z2
            X = X + z2 * s
            f2, df2 = f(X, *args)
            f2 = float(f2)
            df2 = np.array(df2, dtype=float).flatten()
            M = M - 1
            i = i + (1 if length < 0 else 0)
            d2 = np.dot(df2, s)
        
        # End of line search
        
        if success:  # Line search succeeded
            f1 = f2
            fX.append(f1)
            print(f'Iteration {i:4d} | Cost: {f1:4.6e}', end='\r')
            sys.stdout.flush()
            
            # Polack-Ribiere direction
            denom = np.dot(df1, df1)
            if denom != 0:
                s = ((np.dot(df2, df2) - np.dot(df1, df2)) / denom) * s - df2
            else:
                s = -df2
            
            df1 = df2.copy()
            d2 = np.dot(df1, s)
            
            if d2 > 0:  # New slope must be negative
                s = -df1
                d2 = -np.dot(s, s)
            
            z1 = z1 * min(RATIO, d1 / (d2 - np.finfo(float).tiny))
            d1 = d2
            ls_failed = 0
        else:
            X = X0.copy()
            f1 = f0
            df1 = df0.copy()
            
            if ls_failed or i > abs(length):
                break
            
            df1 = df2.copy()
            s = -df1
            d1 = -np.dot(s, s)
            z1 = 1 / (1 - d1)
            ls_failed = 1
    
    print()
    return X, fX, i

File: Linear Regression/fmincg.m
function [X, fX, i] = fmincg(f, X, options, P1, P2, P3, P4, P5)
% Minimize a continuous differentiable multivariate function using a
% conjugate gradient method with line search. The objective function f
% must return both the function value and its gradient.
%
% Usage:
%   [X, fX, i] = fmincg(f, X, options, P1, P2, P3, P4, P5)

% Read optimization options.
if exist('options', 'var') && ~isempty(options) && isfield(options, 'MaxIter')
    maxIter = options.MaxIter;
else
    maxIter = 100;
end

% Line-search constants (Wolfe-Powell conditions).
RHO = 0.01;
SIG = 0.5;
INT = 0.1;
EXT = 3.0;
MAX = 20;
RATIO = 100;

% Build the function-call string for evaluating f with optional parameters.
argstr = ['feval(f, X'];
for idx = 1:(nargin - 3)
    argstr = [argstr, ',P', int2str(idx)];
end
argstr = [argstr, ')'];

if max(size(maxIter)) == 2
    red = maxIter(2);
    maxIter = maxIter(1);
else
    red = 1;
end

statusLabel = 'Iteration ';

% Initialize optimization state.
i = 0;
ls_failed = 0;
fX = [];
[f1, df1] = eval(argstr);          % Evaluate objective and gradient at the initial point.
i = i + (maxIter < 0);

% Start from the steepest descent direction.
s = -df1;
d1 = -s' * s;
z1 = red / (1 - d1);

while i < abs(maxIter)
    i = i + (maxIter > 0);

    % Save current state before attempting a line search.
    X0 = X;
    f0 = f1;
    df0 = df1;

    % Take an initial step along the search direction.
    X = X + z1 * s;
    [f2, df2] = eval(argstr);
    i = i + (maxIter < 0);
    d2 = df2' * s;

    % Initialize the cubic/quadratic interpolation points.
    f3 = f1;
    d3 = d1;
    z3 = -z1;

    if maxIter > 0
        lineSearchMaxEval = MAX;
    else
        lineSearchMaxEval = min(MAX, -maxIter - i);
    end

    success = 0;
    limit = -1;

    while true
        while ((f2 > f1 + z1 * RHO * d1) || (d2 > -SIG * d1)) && (lineSearchMaxEval > 0)
            limit = z1;

            if f2 > f1
                % Quadratic fit.
                z2 = z3 - (0.5 * d3 * z3 * z3) / (d3 * z3 + f2 - f3);
            else
                % Cubic fit.
                A = 6 * (f2 - f3) / z3 + 3 * (d2 + d3);
                B = 3 * (f3 - f2) - z3 * (d3 + 2 * d2);
                z2 = (sqrt(B * B - A * d2 * z3 * z3) - B) / A;
            end

            if isnan(z2) || isinf(z2)
                z2 = z3 / 2;
            end

            z2 = max(min(z2, INT * z3), (1 - INT) * z3);
            z1 = z1 + z2;
            X = X + z2 * s;
            [f2, df2] = eval(argstr);
            lineSearchMaxEval = lineSearchMaxEval - 1;
            i = i + (maxIter < 0);
            d2 = df2' * s;
            z3 = z3 - z2;
        end

        if f2 > f1 + z1 * RHO * d1 || d2 > -SIG * d1
            break;
        elseif d2 > SIG * d1
            success = 1;
            break;
        elseif lineSearchMaxEval == 0
            break;
        end

        % Cubic extrapolation for the next line-search trial.
        A = 6 * (f2 - f3) / z3 + 3 * (d2 + d3);
        B = 3 * (f3 - f2) - z3 * (d3 + 2 * d2);
        z2 = -d2 * z3 * z3 / (B + sqrt(B * B - A * d2 * z3 * z3));

        if ~isreal(z2) || isnan(z2) || isinf(z2) || z2 < 0
            if limit < -0.5
                z2 = z1 * (EXT - 1);
            else
                z2 = (limit - z1) / 2;
            end
        elseif (limit > -0.5) && (z2 + z1 > limit)
            z2 = (limit - z1) / 2;
        elseif (limit < -0.5) && (z2 + z1 > z1 * EXT)
            z2 = z1 * (EXT - 1.0);
        elseif z2 < -z3 * INT
            z2 = -z3 * INT;
        elseif (limit > -0.5) && (z2 < (limit - z1) * (1.0 - INT))
            z2 = (limit - z1) * (1.0 - INT);
        end

        f3 = f2;
        d3 = d2;
        z3 = -z2;
        z1 = z1 + z2;
        X = X + z2 * s;
        [f2, df2] = eval(argstr);
        lineSearchMaxEval = lineSearchMaxEval - 1;
        i = i + (maxIter < 0);
        d2 = df2' * s;
    end

    if success
        f1 = f2;
        fX = [fX' f1]';
        fprintf('%s %4i | Cost: %4.6e\r', statusLabel, i, f1);

        % Polack-Ribiere conjugate gradient update.
        s = (df2' * df2 - df1' * df2) / (df1' * df1) * s - df2;
        tmp = df1;
        df1 = df2;
        df2 = tmp;
        d2 = df1' * s;

        if d2 > 0
            s = -df1;
            d2 = -s' * s;
        end

        z1 = z1 * min(RATIO, d1 / (d2 - realmin));
        d1 = d2;
        ls_failed = 0;
    else
        % Restore the previous state if the line search failed.
        X = X0;
        f1 = f0;
        df1 = df0;

        if ls_failed || i > abs(maxIter)
            break;
        end

        tmp = df1;
        df1 = df2;
        df2 = tmp;
        s = -df1;
        d1 = -s' * s;
        z1 = 1 / (1 - d1);
        ls_failed = 1;
    end

    if exist('OCTAVE_VERSION')
        fflush(stdout);
    end
end

fprintf('\n');
